crates/ra_hir/src/ty/infer/pat.rs

```diff
@@ -0,0 +1,180 @@
+//! Type inference for patterns.
+
+use std::iter::repeat;
+use std::sync::Arc;
+
+use test_utils::tested_by;
+
+use super::{BindingMode, InferenceContext};
+use crate::{
+    db::HirDatabase,
+    expr::{BindingAnnotation, Pat, PatId, RecordFieldPat},
+    ty::{Mutability, Substs, Ty, TypeCtor, TypeWalk},
+    Name, Path,
+};
+
+impl<'a, D: HirDatabase> InferenceContext<'a, D> {
+    fn infer_tuple_struct_pat(
+        &mut self,
+        path: Option<&Path>,
+        subpats: &[PatId],
+        expected: &Ty,
+        default_bm: BindingMode,
+    ) -> Ty {
+        let (ty, def) = self.resolve_variant(path);
+
+        self.unify(&ty, expected);
+
+        let substs = ty.substs().unwrap_or_else(Substs::empty);
+
+        for (i, &subpat) in subpats.iter().enumerate() {
+            let expected_ty = def
+                .and_then(|d| d.field(self.db, &Name::new_tuple_field(i)))
+                .map_or(Ty::Unknown, |field| field.ty(self.db))
+                .subst(&substs);
+            let expected_ty = self.normalize_associated_types_in(expected_ty);
+            self.infer_pat(subpat, &expected_ty, default_bm);
+        }
+
+        ty
+    }
+
+    fn infer_record_pat(
+        &mut self,
+        path: Option<&Path>,
+        subpats: &[RecordFieldPat],
+        expected: &Ty,
+        default_bm: BindingMode,
+        id: PatId,
+    ) -> Ty {
+        let (ty, def) = self.resolve_variant(path);
+        if let Some(variant) = def {
+            self.write_variant_resolution(id.into(), variant);
+        }
+
+        self.unify(&ty, expected);
+
+        let substs = ty.substs().unwrap_or_else(Substs::empty);
+
+        for subpat in subpats {
+            let matching_field = def.and_then(|it| it.field(self.db, &subpat.name));
+            let expected_ty =
+                matching_field.map_or(Ty::Unknown, |field| field.ty(self.db)).subst(&substs);
+            let expected_ty = self.normalize_associated_types_in(expected_ty);
+            self.infer_pat(subpat.pat, &expected_ty, default_bm);
+        }
+
+        ty
+    }
+
+    pub(super) fn infer_pat(
```

Comment:
I slightly prefer to keep public functions at the top. I've got a mild case of ptsd after encountering the rustc parser for the first time. I had to plow via 6k lines of rust to find an entry point at the very end :)